### ISSUE TITLE
Imagesize support for Structured Data (JSON-LD) and Open Graph

### ISSUE THREAD

**Issue by AileenCGN** [TryGhost/Ghost#7095]
To improve our structured data, we need to change all our image properties to be `ImageObjects`. This means, we need to add a `width` and `height` property to all of them. To get the dimensions of an image, I want to implement the npm module `[image-size](https://github.com/image-size/image-size)` which will get the necessary dimensions on the fly.

---

## Schema.org (JSON-LD)
This is our status quo (for example `publisher`):

```javascript
"@context": "https://schema.org",
"@type": "Article",
"publisher": {
    "@type": "Organization",
    "name": "My Ghost Blog",
    "logo": "http://myblog.com/content/images/2016/06/IMG_1234.jpg"
    }
```

And this is how it should look like:

```javascript
"@context": "https://schema.org",
"@type": "Article",
"publisher": {
    "@type": "Organization",
    "name": "My Ghost Blog",
    "logo": {
        "@type": "ImageObject",
        "url": "http://myblog.com/content/images/2016/06/IMG_1234.jpg"
        "width": 600,
        "height": 60
        }
    }
```
**Note:** Google has [restrictions](https://developers.google.com/search/docs/data-types/articles) on the image dimension, especially for the publisher logo. The image needs to be <=600px wide and <=60px high (ideally exactly 600px x 60px). Unless we have proper image-handling (see #4453), we will not output an `ImageObject` if the logo doesn't fit in the dimensions.

**Applies to**
- Schema for _Post/Page_:
	- `publisher.logo`
	-  `author.image`
	-  `image`
- Schema for _Home/Website_:
	- `publisher.logo`
	- `image`
- Schema for _Tag_:
	- `publisher.logo`
	- `image`
- Schema for _Author_:
	- `image`  

---

Add `mainEntityOfPage` to `JSON-LD` as it is now required by Google:

```javascript
"mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "http://myblog.com/",
    }
```
**Applies to**
- Schema for _Post/Page_
- Schema for _Home/Website_
- Schema for _Tag_
- Schema for _Author_

---

## Open Graph (Facebook)

Now, we have the dimensions of images, we can improve our open graph data as well by adding

- `og:image:width` and
- `og:image:height`

---
ToDos:
- [x] Add `image-size` to dependencies
- [x] Make `getImageSize` util
- [x] Include restrictions for `publisher.logo` (<=600px width and <=60px height)
- [x] Update `image` properties in structured data to be full `ImageObjects`
- [x] Add `mainEntityOfPage` to schema.org
- [x] Add new properties `og:image:width` and `og:image:height` to structured data
- [x] Use cache for image dimensions
- [x] Write/update tests

**Comment by halfdan** [comment]
Two things here:

- How will this work with images that are stored externally? (S3, CloudFront, ...)
- This [issue](https://github.com/image-size/image-size/issues/76) is a bit concerning. Often people upload pictures that are rotated client side based on EXIF data. This would mean that these pictures incorrectly have their width/height swapped.

**Comment by AileenCGN** [comment]
>How will this work with images that are stored externally? (S3, CloudFront, ...)

The `image-size` module fetches the image dimensions from a URL, as well as locally stored images.

>This issue is a bit concerning. Often people upload pictures that are rotated client side based on EXIF data. This would mean that these pictures incorrectly have their width/height swapped.

I'm aware, that we still have this issue with the EXIF data. This feature will be an v1 improvement until we have proper image handling. We will need to get the image size now anyways as we want to have AMP support, where we have to provide the `width` and `height` for _every_ image in a blog post.

**Comment by halfdan** [comment]
> The image-size module fetches the image dimensions from a URL, as well as locally stored images.

That's really cool! Will there be any caching done on the server to avoid having to fetch images over and over again?

**Comment by AileenCGN** [comment]
This is merged to `amp` branch, but will stay open until it's merged to `master`.

**Comment by kirrg001** [comment]
implemented in #7229

**Comment by puzrin** [comment]
I"d suggest to use https://github.com/nodeca/probe-image-size. It supports partial download for remote images, svg dimention units, and has none or errors reported in image-size.
